Architecture updater (auto-sync) - Updating PPC

Rot127 opened the pull request:
Base on: https://github.com/capstone-engine/capstone/pull/1949

Draft PR for the `auto-sync` refactor of PPC.

It will also refactor quite a lot of general mapping logic used by `auto-sync` archs

**TODO**

- [x] PS instructions
- [x] Make Capstone build with generated and translated files.
- [x] Add details
  - [x] Print different deatil for branch hints and add `not set` flag.
  - [x] Move `add_cs_detail` to new `cs_mapping.h`
  - [x] Implement operand handling
  - [x] Handle `TH` operands for `dcbt`
  - [x] Fix `td` files
- [x] `"0x04,0x00,0x00,0x00,0xf4,0xa4,0x00,0x08"` -> `pstd 5, 8(4), 0`
- [x] Fuzz more
- [x] Update and add tests

**Optimizations to do afterwards** (in another PR)

- [x] Macros should end with `;` for proper formatting.
- [ ] Data types of mapped operands should be excluded for `CAPSTONE_DIET`
- [x] Add `get_cs_reg_alias()` for non LLVM register names.
- [ ] Add `add_cs_group()` for non LLVM instruction groups.
- [ ] Add asserts (again for translated files) which are disabled by default.
- [ ] Upstream additions in `PPCPredicate.h` to LLVM

**Closes**

closes https://github.com/capstone-engine/capstone/issues/1909
closes https://github.com/capstone-engine/capstone/issues/476
closes https://github.com/capstone-engine/capstone/issues/1936
closes https://github.com/capstone-engine/capstone/issues/1914
closes https://github.com/capstone-engine/capstone/issues/1912
closes https://github.com/capstone-engine/capstone/issues/1693
closes https://github.com/capstone-engine/capstone/issues/1527
closes https://github.com/capstone-engine/capstone/issues/944

**Test again**

- https://github.com/capstone-engine/capstone/issues/968 (ppc64be hits assert)
- https://github.com/capstone-engine/capstone/issues/969 (hints are broken)
- https://github.com/capstone-engine/capstone/issues/970 (hints broken)
- https://github.com/capstone-engine/capstone/issues/1914 (missing reg alias)
- https://github.com/capstone-engine/capstone/issues/1746 (What does LLVM)
- https://github.com/capstone-engine/capstone/issues/1527 (missing operand)
- https://github.com/capstone-engine/capstone/issues/1528 (What does LLVM)
- https://github.com/capstone-engine/capstone/issues/1117 (no reg alias, what does LLVM?)
- https://github.com/capstone-engine/capstone/issues/1903 (hits assert)
- https://github.com/capstone-engine/capstone/issues/1596 (can't build fuzzer currently)


**Breaking changes**
- Predicate enums members are renamed. They now use the LLVM name (e.g. `PPC_BC_NU_PLUS` -> `PPC_PRED_NU_PLUS`).
- Branch conditions are now saved in more detail in `cs_ppc.bc`.
- The base register of an PPC memory operand was not present if `reg = r0`. This is fixed now.
- `ppc_ops_crx` is removed (wasn't used).

Rot127 commented:
Hi @terorie! As you see I am just in the process to refactor the PPC module. You added the `paired-single` instructions, but I couldn't find them in LLVM. Have you used this revision (https://reviews.llvm.org/D85137)? Or were the `PS` instructions at some point part of LLVM and they are now deprecated? Could you tell me the release you used in the latter case?

riptl commented:
Hey @Rot127 unfortunately I made minimal progress here.

There was a TableGen version mismatch between DarkKirb's patch and the one that Capstone needed. 

So, I started learning and rewriting TableGen rules by hand a long time ago but not sure where I left that off.

Will share the code if I can find it again. Happy to pick this back up if there's interest.

Rot127 commented:
> Will share the code if I can find it again.

This would be great!

> Happy to pick this back up if there's interest.

While I am at it I would try to add it on my own. But can I come back to your offer if I get stuck? The priority is to focus on refactoring as many archs as possible to use `auto-sync`

XVilka commented on arch/PowerPC/PPCDisassembler.c:
```diff
@@ -1,627 +1,515 @@
+/* LLVM-tag: <tag> */
+
+/* Only small edits allowed. */
+/* For multiple similiar edits, please create a Patch for the translator. */
```
A typo, "similiar" should be "similar"

XVilka commented on arch/PowerPC/PPCDisassembler.c:
```diff
@@ -1,627 +1,515 @@
+}
 
+static DecodeStatus decodeMemRIHashOperands(MCInst *Inst, uint64_t Imm,
+											int64_t Address,
```
Same problem of huge empty spaces?

Rot127 replied:
@XVilka This depends on the tab width in Github or your editor. Github defaults to 8 spaces per tab.

XVilka reviewed (changes requested):
Two minor things

Rot127 commented:
@kabeor @XVilka What are your opinions regarding `asserts`? They were removed in Capstone (for performance I guess?) and so I removed them as well during translation. Should they be reintroduced? But wrapped in a macro. So we can only use it for debugging?

XVilka commented:
@Rot127 no objections against asserts if they are disabled by default. Maybe could be something like `cs_assert()` to avoid any possible clash with the standard one?

Rot127 commented:
@terorie The `PS` instructions are now implemented. Wasn't too much trouble.

<details> 
  <summary>Details</summary>

  ```
  ./cstool -d ppc32be "\x10\x00\x1f\xec\xe0\x6d\x80\x04\xe4\x6d\x80\x04\x10\x60\x1c\x4c\x10\x60\x1c\x0c\xf0\x6d\x80\x04\xf4\x6d\x80\x04\x10\x60\x1c\x4e\x10\x60\x1c\x0e\x10\x60\x1a\x10\x10\x60\x1a\x11\x10\x63\x20\x2a\x10\x63\x20\x2b\x10\x83\x20\x40\x10\x83\x20\xC0\x10\x83\x20\x00\x10\x83\x20\x80\x10\x63\x20\x24\x10\x63\x20\x25\x10\x63\x29\x3a\x10\x63\x29\x3b\x10\x63\x29\x1c\x10\x63\x29\x1d\x10\x63\x29\x1e\x10\x63\x29\x1f\x10\x63\x24\x20\x10\x63\x24\x21\x10\x63\x24\x60\x10\x63\x24\x61\x10\x63\x24\xA0\x10\x63\x24\xA1\x10\x63\x24\xE0\x10\x63\x24\xE1\x10\x60\x20\x90\x10\x60\x20\x91\x10\x63\x29\x38\x10\x63\x29\x39\x10\x63\x01\x32\x10\x63\x01\x33\x10\x63\x01\x18\x10\x63\x01\x19\x10\x63\x01\x1A\x10\x63\x01\x1B\x10\x60\x19\x10\x10\x60\x19\x11\x10\x60\x18\x50\x10\x60\x18\x51\x10\x63\x29\x3e\x10\x63\x29\x3f\x10\x63\x29\x3c\x10\x63\x29\x3d\x10\x60\x18\x30\x10\x60\x18\x31\x10\x60\x18\x34\x10\x60\x18\x35\x10\x63\x29\x2e\x10\x63\x29\x2f\x10\x63\x20\x28\x10\x63\x20\x29\x10\x63\x29\x14\x10\x63\x29\x15\x10\x63\x29\x16\x10\x63\x29\x17"
   0  10 00 1f ec  dcbz_l	r0, r3
	  ID: 730 (popcntw)
	  op_count: 2
		  operands[0].type: REG = r0
		  operands[0].access: READ
		  operands[1].type: REG = r3
		  operands[1].access: READ
	  Groups: HasPS 
  
   4  e0 6d 80 04  psq_l	f3, 4(r13), 1, 0
	  ID: 731 (dcbz_l)
	  op_count: 4
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: MEM
			  operands[1].mem.base: REG = r13
			  operands[1].mem.disp: 0x4
		  operands[1].access: READ
		  operands[2].type: IMM = 0x1
		  operands[2].access: READ
		  operands[3].type: IMM = 0x0
		  operands[3].access: READ
	  Groups: HasPS 
  
   8  e4 6d 80 04  psq_lu	f3, 4(r13), 1, 0
	  ID: 732 (psq_l)
	  op_count: 4
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: MEM
			  operands[1].mem.base: REG = r13
			  operands[1].mem.disp: 0x4
		  operands[1].access: READ
		  operands[2].type: IMM = 0x1
		  operands[2].access: READ
		  operands[3].type: IMM = 0x0
		  operands[3].access: READ
	  Groups: HasPS 
  
   c  10 60 1c 4c  psq_lux	f3, r0, r3, 1, 0
	  ID: 733 (psq_lu)
	  op_count: 5
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = r0
		  operands[1].access: READ
		  operands[2].type: REG = r3
		  operands[2].access: READ
		  operands[3].type: IMM = 0x1
		  operands[3].access: READ
		  operands[4].type: IMM = 0x0
		  operands[4].access: READ
	  Groups: HasPS 
  
  10  10 60 1c 0c  psq_lx	f3, r0, r3, 1, 0
	  ID: 734 (psq_lux)
	  op_count: 5
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = r0
		  operands[1].access: READ
		  operands[2].type: REG = r3
		  operands[2].access: READ
		  operands[3].type: IMM = 0x1
		  operands[3].access: READ
		  operands[4].type: IMM = 0x0
		  operands[4].access: READ
	  Groups: HasPS 
  
  14  f0 6d 80 04  psq_st	f3, 4(r13), 1, 0
	  ID: 735 (psq_lx)
	  op_count: 4
		  operands[0].type: REG = f3
		  operands[0].access: READ
		  operands[1].type: MEM
			  operands[1].mem.base: REG = r13
			  operands[1].mem.disp: 0x4
		  operands[1].access: WRITE
		  operands[2].type: IMM = 0x1
		  operands[2].access: READ
		  operands[3].type: IMM = 0x0
		  operands[3].access: READ
	  Groups: HasPS 
  
  18  f4 6d 80 04  psq_stu	f3, 4(r13), 1, 0
	  ID: 736 (psq_st)
	  op_count: 4
		  operands[0].type: REG = f3
		  operands[0].access: READ
		  operands[1].type: MEM
			  operands[1].mem.base: REG = r13
			  operands[1].mem.disp: 0x4
		  operands[1].access: WRITE
		  operands[2].type: IMM = 0x1
		  operands[2].access: READ
		  operands[3].type: IMM = 0x0
		  operands[3].access: READ
	  Groups: HasPS 
  
  1c  10 60 1c 4e  psq_stux	f3, r0, r3, 1, 0
	  ID: 737 (psq_stu)
	  op_count: 5
		  operands[0].type: REG = f3
		  operands[0].access: READ
		  operands[1].type: REG = r0
		  operands[1].access: READ
		  operands[2].type: REG = r3
		  operands[2].access: READ
		  operands[3].type: IMM = 0x1
		  operands[3].access: READ
		  operands[4].type: IMM = 0x0
		  operands[4].access: READ
	  Groups: HasPS 
  
  20  10 60 1c 0e  psq_stx	f3, r0, r3, 1, 0
	  ID: 738 (psq_stux)
	  op_count: 5
		  operands[0].type: REG = f3
		  operands[0].access: READ
		  operands[1].type: REG = r0
		  operands[1].access: READ
		  operands[2].type: REG = r3
		  operands[2].access: READ
		  operands[3].type: IMM = 0x1
		  operands[3].access: READ
		  operands[4].type: IMM = 0x0
		  operands[4].access: READ
	  Groups: HasPS 
  
  24  10 60 1a 10  ps_abs	f3, f3
	  ID: 749 (pstxvp)
	  op_count: 2
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
	  Groups: HasPS 
  
  28  10 60 1a 11  ps_abs.	f3, f3
	  ID: 749 (pstxvp)
	  op_count: 2
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
	  Groups: HasPS 
  
  2c  10 63 20 2a  ps_add	f3, f3, f4
	  ID: 750 (ps_abs)
	  op_count: 3
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
	  Groups: HasPS 
  
  30  10 63 20 2b  ps_add.	f3, f3, f4
	  ID: 750 (ps_abs)
	  op_count: 3
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
	  Groups: HasPS 
  
  34  10 83 20 40  ps_cmpo0	cr1, f3, f4
	  ID: 751 (ps_add)
	  op_count: 3
		  operands[0].type: REG = cr1
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
	  Groups: HasPS 
  
  38  10 83 20 c0  ps_cmpo1	cr1, f3, f4
	  ID: 752 (ps_cmpo0)
	  op_count: 3
		  operands[0].type: REG = cr1
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
	  Groups: HasPS 
  
  3c  10 83 20 00  ps_cmpu0	cr1, f3, f4
	  ID: 753 (ps_cmpo1)
	  op_count: 3
		  operands[0].type: REG = cr1
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
	  Groups: HasPS 
  
  40  10 83 20 80  ps_cmpu1	cr1, f3, f4
	  ID: 754 (ps_cmpu0)
	  op_count: 3
		  operands[0].type: REG = cr1
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
	  Groups: HasPS 
  
  44  10 63 20 24  ps_div	f3, f3, f4
	  ID: 755 (ps_cmpu1)
	  op_count: 3
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
	  Groups: HasPS 
  
  48  10 63 20 25  ps_div.	f3, f3, f4
	  ID: 755 (ps_cmpu1)
	  op_count: 3
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
	  Groups: HasPS 
  
  4c  10 63 29 3a  ps_madd	f3, f3, f4, f5
	  ID: 756 (ps_div)
	  op_count: 4
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
		  operands[3].type: REG = f5
		  operands[3].access: READ
	  Groups: HasPS 
  
  50  10 63 29 3b  ps_madd.	f3, f3, f4, f5
	  ID: 756 (ps_div)
	  op_count: 4
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
		  operands[3].type: REG = f5
		  operands[3].access: READ
	  Groups: HasPS 
  
  54  10 63 29 1c  ps_madds0	f3, f3, f4, f5
	  ID: 757 (ps_madd)
	  op_count: 4
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
		  operands[3].type: REG = f5
		  operands[3].access: READ
	  Groups: HasPS 
  
  58  10 63 29 1d  ps_madds0.	f3, f3, f4, f5
	  ID: 757 (ps_madd)
	  op_count: 4
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
		  operands[3].type: REG = f5
		  operands[3].access: READ
	  Groups: HasPS 
  
  5c  10 63 29 1e  ps_madds1	f3, f3, f4, f5
	  ID: 758 (ps_madds0)
	  op_count: 4
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
		  operands[3].type: REG = f5
		  operands[3].access: READ
	  Groups: HasPS 
  
  60  10 63 29 1f  ps_madds1.	f3, f3, f4, f5
	  ID: 758 (ps_madds0)
	  op_count: 4
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
		  operands[3].type: REG = f5
		  operands[3].access: READ
	  Groups: HasPS 
  
  64  10 63 24 20  ps_merge00	f3, f3, f4
	  ID: 759 (ps_madds1)
	  op_count: 3
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
	  Groups: HasPS 
  
  68  10 63 24 21  ps_merge00.	f3, f3, f4
	  ID: 759 (ps_madds1)
	  op_count: 3
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
	  Groups: HasPS 
  
  6c  10 63 24 60  ps_merge01	f3, f3, f4
	  ID: 760 (ps_merge00)
	  op_count: 3
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
	  Groups: HasPS 
  
  70  10 63 24 61  ps_merge01.	f3, f3, f4
	  ID: 760 (ps_merge00)
	  op_count: 3
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
	  Groups: HasPS 
  
  74  10 63 24 a0  ps_merge10	f3, f3, f4
	  ID: 761 (ps_merge01)
	  op_count: 3
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
	  Groups: HasPS 
  
  78  10 63 24 a1  ps_merge10.	f3, f3, f4
	  ID: 761 (ps_merge01)
	  op_count: 3
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
	  Groups: HasPS 
  
  7c  10 63 24 e0  ps_merge11	f3, f3, f4
	  ID: 762 (ps_merge10)
	  op_count: 3
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
	  Groups: HasPS 
  
  80  10 63 24 e1  ps_merge11.	f3, f3, f4
	  ID: 762 (ps_merge10)
	  op_count: 3
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
	  Groups: HasPS 
  
  84  10 60 20 90  ps_mr	f3, f4
	  ID: 763 (ps_merge11)
	  op_count: 2
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f4
		  operands[1].access: READ
	  Groups: HasPS 
  
  88  10 60 20 91  ps_mr.	f3, f4
	  ID: 763 (ps_merge11)
	  op_count: 2
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f4
		  operands[1].access: READ
	  Groups: HasPS 
  
  8c  10 63 29 38  ps_msub	f3, f3, f4, f5
	  ID: 764 (ps_mr)
	  op_count: 4
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
		  operands[3].type: REG = f5
		  operands[3].access: READ
	  Groups: HasPS 
  
  90  10 63 29 39  ps_msub.	f3, f3, f4, f5
	  ID: 764 (ps_mr)
	  op_count: 4
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
		  operands[3].type: REG = f5
		  operands[3].access: READ
	  Groups: HasPS 
  
  94  10 63 01 32  ps_mul	f3, f3, f4
	  ID: 765 (ps_msub)
	  op_count: 3
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
	  Groups: HasPS 
  
  98  10 63 01 33  ps_mul.	f3, f3, f4
	  ID: 765 (ps_msub)
	  op_count: 3
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
	  Groups: HasPS 
  
  9c  10 63 01 18  ps_muls0	f3, f3, f4
	  ID: 766 (ps_mul)
	  op_count: 3
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
	  Groups: HasPS 
  
  a0  10 63 01 19  ps_muls0.	f3, f3, f4
	  ID: 766 (ps_mul)
	  op_count: 3
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
	  Groups: HasPS 
  
  a4  10 63 01 1a  ps_muls1	f3, f3, f4
	  ID: 767 (ps_muls0)
	  op_count: 3
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
	  Groups: HasPS 
  
  a8  10 63 01 1b  ps_muls1.	f3, f3, f4
	  ID: 767 (ps_muls0)
	  op_count: 3
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
	  Groups: HasPS 
  
  ac  10 60 19 10  ps_nabs	f3, f3
	  ID: 768 (ps_muls1)
	  op_count: 2
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
	  Groups: HasPS 
  
  b0  10 60 19 11  ps_nabs.	f3, f3
	  ID: 768 (ps_muls1)
	  op_count: 2
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
	  Groups: HasPS 
  
  b4  10 60 18 50  ps_neg	f3, f3
	  ID: 769 (ps_nabs)
	  op_count: 2
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
	  Groups: HasPS 
  
  b8  10 60 18 51  ps_neg.	f3, f3
	  ID: 769 (ps_nabs)
	  op_count: 2
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
	  Groups: HasPS 
  
  bc  10 63 29 3e  ps_nmadd	f3, f3, f4, f5
	  ID: 770 (ps_neg)
	  op_count: 4
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
		  operands[3].type: REG = f5
		  operands[3].access: READ
	  Groups: HasPS 
  
  c0  10 63 29 3f  ps_nmadd.	f3, f3, f4, f5
	  ID: 770 (ps_neg)
	  op_count: 4
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
		  operands[3].type: REG = f5
		  operands[3].access: READ
	  Groups: HasPS 
  
  c4  10 63 29 3c  ps_nmsub	f3, f3, f4, f5
	  ID: 771 (ps_nmadd)
	  op_count: 4
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
		  operands[3].type: REG = f5
		  operands[3].access: READ
	  Groups: HasPS 
  
  c8  10 63 29 3d  ps_nmsub.	f3, f3, f4, f5
	  ID: 771 (ps_nmadd)
	  op_count: 4
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
		  operands[3].type: REG = f5
		  operands[3].access: READ
	  Groups: HasPS 
  
  cc  10 60 18 30  ps_res	f3, f3
	  ID: 772 (ps_nmsub)
	  op_count: 2
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
	  Groups: HasPS 
  
  d0  10 60 18 31  ps_res.	f3, f3
	  ID: 772 (ps_nmsub)
	  op_count: 2
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
	  Groups: HasPS 
  
  d4  10 60 18 34  ps_rsqrte	f3, f3
	  ID: 773 (ps_res)
	  op_count: 2
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
	  Groups: HasPS 
  
  d8  10 60 18 35  ps_rsqrte.	f3, f3
	  ID: 773 (ps_res)
	  op_count: 2
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
	  Groups: HasPS 
  
  dc  10 63 29 2e  ps_sel	f3, f3, f4, f5
	  ID: 774 (ps_rsqrte)
	  op_count: 4
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
		  operands[3].type: REG = f5
		  operands[3].access: READ
	  Groups: HasPS 
  
  e0  10 63 29 2f  ps_sel.	f3, f3, f4, f5
	  ID: 774 (ps_rsqrte)
	  op_count: 4
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
		  operands[3].type: REG = f5
		  operands[3].access: READ
	  Groups: HasPS 
  
  e4  10 63 20 28  ps_sub	f3, f3, f4
	  ID: 775 (ps_sel)
	  op_count: 3
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
	  Groups: HasPS 
  
  e8  10 63 20 29  ps_sub.	f3, f3, f4
	  ID: 775 (ps_sel)
	  op_count: 3
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
	  Groups: HasPS 
  
  ec  10 63 29 14  ps_sum0	f3, f3, f4, f5
	  ID: 776 (ps_sub)
	  op_count: 4
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
		  operands[3].type: REG = f5
		  operands[3].access: READ
	  Groups: HasPS 
  
  f0  10 63 29 15  ps_sum0.	f3, f3, f4, f5
	  ID: 776 (ps_sub)
	  op_count: 4
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
		  operands[3].type: REG = f5
		  operands[3].access: READ
	  Groups: HasPS 
  
  f4  10 63 29 16  ps_sum1	f3, f3, f4, f5
	  ID: 777 (ps_sum0)
	  op_count: 4
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
		  operands[3].type: REG = f5
		  operands[3].access: READ
	  Groups: HasPS 
  
  f8  10 63 29 17  ps_sum1.	f3, f3, f4, f5
	  ID: 777 (ps_sum0)
	  op_count: 4
		  operands[0].type: REG = f3
		  operands[0].access: WRITE
		  operands[1].type: REG = f3
		  operands[1].access: READ
		  operands[2].type: REG = f4
		  operands[2].access: READ
		  operands[3].type: REG = f5
		  operands[3].access: READ
	  Groups: HasPS
  ```

</detail>

Rot127 committed:
Refactor PPC module for auto-sync

Rot127 committed:
Update auto-sync scripts with changes introdduces with PPC update

Rot127 committed:
Update additional files associated with PPC auto-sync update.

Rot127 commented:
Rebased and squashed this PR onto the latest commit of https://github.com/capstone-engine/capstone/pull/1949 (683a5951c7b995482766f5936117a51bc59e769d).
Original commit history is here: https://github.com/rot127/capstone/tree/auto-sync-ppc-bak

XVilka commented:
@Rot127 some fuzzer job bugs found - check the CI: https://github.com/capstone-engine/capstone/actions/runs/5494173005/jobs/10012800208?pr=2013#step:5:82

Rot127 commented:
Yes, I am going to fuzz it as well before starting with Rizin integration.

Rot127 committed:
Add check for detail

Rot127 committed:
Add getFeatureBits call.

Rot127 committed:
Fix reachable assert/segfault (also in LLVM).

Rot127 committed:
Remove INVALID entry which is already part of the generated data

Rot127 committed:
Handle memory operands printed via printOperand without setting mem op before.

Rot127 committed:
Set correct operand type of cr bit field.

Rot127 committed:
Remove debug string

Rot127 committed:
Fix braces around initializer bug

Rot127 committed:
Add naive fuzzing code.

Rot127 committed:
Add MT/MFSPR alias enum values

Rot127 committed:
Rename duplicate variable

Rot127 committed:
Revert "Rename duplicate variable"

This reverts commit 982202cf1a7092a62122b6946d35b608cabc078b.

Rot127 committed:
Revert "Add naive fuzzing code."

This reverts commit ea9660ea382def6577acc2b4f12e35603bc28e41.

XVilka commented:
Please rebase on top of latest `next`

Rot127 committed:
Add missing operand printers

Rot127 committed:
Fix undefined behavior of shift by using uint64_t and fix build warning of unused function.

Rot127 committed:
Add more MTSPR and MFSPR alias

Rot127 committed:
Increment major version

XVilka commented on include/capstone/capstone.h:
```diff
@@ -48,7 +48,7 @@ extern "C" {
 
 // Capstone API version
-#define CS_API_MAJOR 5
+#define CS_API_MAJOR 6
```
Please send this as a separate PR, to merge first

Rot127 replied:
Yes. Extract it after the 0.6 release.

XVilka replied:
@Rot127 I think it's better to be done *before* the release, to clearly denote that the `next` branch is different from the `v5` one. @kabeor @aquynh what do you think?

Rot127 replied:
Ah, my bad. I thought this was the Rizin PR. Too many PRs...

Rot127 replied:
https://github.com/capstone-engine/capstone/pull/2104

kabeor replied:
I think it's better to change this as Major 5 Minor 2

Rot127 committed:
Fix rebase blunder. Set correct header file location.

Rot127 committed:
Allow to patch also inc files without specifing the include guard.

Rot127 committed:
Fix: Cutting first line.

Rot127 committed:
Add feature to select between alias and real operand set.

Alias instructions always had the operand set of the alias instruction.
This can hide certain important operands.
This commit adds the possibility to print the alias instructions as always,
but have the detail operands set as if it would be the real instructions.

This greatly helps increases the consistent use of detail->operands.

Rot127 committed:
Clearify documentation and fix incorrect SO predicates.

Rot127 committed:
Save BI field separately from csX_bit.

Rot127 committed:
Replace magic numbers with their enums.

Rot127 committed:
Add helper functions to test for branch cond flags.

Rot127 committed:
Remove LI instruction (they are alias).

Rot127 committed:
Set constant zero as mem base if base reg == r0.

Rot127 committed:
Replace tabs after mnemonic with space.

Rot127 committed:
Set ZERO register, not the value 0 for (RA|0) cases

Rot127 committed:
Add missing implicit write to LR.

Rot127 committed:
Print implicity read/written registers.

Rot127 committed:
Remove unused code (was replaced with proper alias handling.)

Rot127 committed:
Fix S[LR]WI alias instructions.

Rot127 committed:
Handle hardcoded sldi alias.

Rot127 committed:
Remove unused variable

Rot127 committed:
Remove leading tabs

Rot127 committed:
Check if detail ops shall be filled before insert_op

Rot127 committed:
Replace branch alias defined as instructions with actual alias.

Rot127 committed:
Remove additional space before alias mnemonic

Rot127 committed:
Rename function to make it less similar to test_cr_bit and add harder check for true

Rot127 committed:
Fix build of cstest ppc_detail.c

Rot127 committed:
Reorder alias after change to td files.

Rot127 committed:
Set correct type instruction decoding (fixes prefixed instructions.)

Rot127 committed:
Add memory operand flags to prefix instructions.

Rot127 committed:
Check for option to print register names without prefix.

Rot127 committed:
Add check if PS mode is enabled.

Rot127 commented:
@kabeor @aquynh So this one is pretty much ready for review. A handful tests still fail, but I'll fix them tomorrow.

Rot127 committed:
Revert "Increment major version"

This reverts commit 659e7da80446b580eebe75afb6f71a2906edfbe6.

XVilka commented:
@wargio as a PPC expert please take a look when you have time too.

Rot127 committed:
Add READ_WRITE access attribute to fix build error.

Rot127 committed:
Run clang-format

XVilka commented on Mapping.c:
```diff
@@ -311,3 +312,69 @@ const cs_ac_type mapping_get_op_access(MCInst *MI, unsigned OpNum,
+
+/// Returns true if for this architecture the
+/// alias operands should be filled.
+/// TODO: Replace this with a proper option.
```
Do we have an issue for this?

Rot127 replied:
Not yet. I need to update `AArch64` and `ARM` for it as well. Do this after or just before everything is updated to LLVM 17.

XVilka commented on Mapping.c:
```diff
@@ -311,3 +312,69 @@ const cs_ac_type mapping_get_op_access(MCInst *MI, unsigned OpNum,
+}
+
+/// Sets the setDetailOps flag to @p Val.
+/// If detail == NULLit refuses to set the flag to true.
```
typo?

XVilka commented on suite/auto-sync/Update-Arch.sh:
```diff
@@ -53,7 +53,7 @@ setup_build_dir() {
 #
 
-supported="ARM"
+supported="ARM, PPC"
```
After ARM64 it's probably better to convert this script to Python too

Rot127 replied:
Yes. Start with a script which updates the `inc` files today (it also applies the patches and updates the commit/tag note). Then, convert it before LLVM 17 rebase. This way we can test it directly.

XVilka reviewed (approved):
The code is in excellent shape. Fantastic work, as usual! LGTM as is.

Rot127 commented:
@kabeor After https://github.com/capstone-engine/capstone/pull/2126 is merged I update the tests here. Since we already tested a lot in Rizin there should be minimal fixes/additions if at all.

XVilka commented:
> @kabeor After #2126 is merged I update the tests here. Since we already tested a lot in Rizin there should be minimal fixes/additions if at all.

@Rot127 have you had a chance to do that? I think it's the only missing piece that prevents this to be merged (in my opinion).

XVilka commented:
Failed tests:
```
[-] There are/is 1 failed test(s)
[-] Detailed report for ../MC/PowerPC/ppc64-encoding-bookII.s.cs:

	[+] Line 16: 0x7c,0x00,0x04,0xac
		"sync 0" != "sync" ( "sync 0" != "sync" and "sync 0" != "sync" )

[-] Target: ppc64-operands.s.cs

[-] There are/is 13 failed test(s)
[-] Detailed report for ../MC/PowerPC/ppc64-encoding.s.cs:

	[+] Line 10: 0x4c,0x43,0x22,0x02
		"crand eq, un, 4*cr1+lt" != "crand cr0eq, cr0un, cr1lt" ( "crand eq, un, 4*cr1+lt" != "crand cr0eq, cr0un, cr1lt" and "crand 2, 3, 4" != "crand cr0eq, cr0un, cr1lt" )

	[+] Line 11: 0x4c,0x43,0x21,0xc2
		"crnand 2, 3, 4" != "crnand cr0eq, cr0un, cr1lt" ( "crnand 2, 3, 4" != "crnand cr0eq, cr0un, cr1lt" and "crnand 2, 3, 4" != "crnand cr0eq, cr0un, cr1lt" )

	[+] Line 12: 0x4c,0x43,0x23,0x82
		"cror 2, 3, 4" != "cror cr0eq, cr0un, cr1lt" ( "cror 2, 3, 4" != "cror cr0eq, cr0un, cr1lt" and "cror 2, 3, 4" != "cror cr0eq, cr0un, cr1lt" )

	[+] Line 13: 0x4c,0x43,0x21,0x82
		"crxor 2, 3, 4" != "crxor cr0eq, cr0un, cr1lt" ( "crxor 2, 3, 4" != "crxor cr0eq, cr0un, cr1lt" and "crxor 2, 3, 4" != "crxor cr0eq, cr0un, cr1lt" )

	[+] Line 14: 0x4c,0x43,0x20,0x42
		"crnor 2, 3, 4" != "crnor cr0eq, cr0un, cr1lt" ( "crnor 2, 3, 4" != "crnor cr0eq, cr0un, cr1lt" and "crnor 2, 3, 4" != "crnor cr0eq, cr0un, cr1lt" )

	[+] Line 15: 0x4c,0x43,0x22,0x42
		"creqv 2, 3, 4" != "creqv cr0eq, cr0un, cr1lt" ( "creqv 2, 3, 4" != "creqv cr0eq, cr0un, cr1lt" and "creqv 2, 3, 4" != "creqv cr0eq, cr0un, cr1lt" )

	[+] Line 16: 0x4c,0x43,0x21,0x02
		"crandc 2, 3, 4" != "crandc cr0eq, cr0un, cr1lt" ( "crandc 2, 3, 4" != "crandc cr0eq, cr0un, cr1lt" and "crandc 2, 3, 4" != "crandc cr0eq, cr0un, cr1lt" )

	[+] Line 17: 0x4c,0x43,0x23,0x42
		"crorc 2, 3, 4" != "crorc cr0eq, cr0un, cr1lt" ( "crorc 2, 3, 4" != "crorc cr0eq, cr0un, cr1lt" and "crorc 2, 3, 4" != "crorc cr0eq, cr0un, cr1lt" )

	[+] Line 72: 0x7c,0x43,0x20,0x50
		"sub %r2, %r4, %r3" != "subf 2, 3, 4" ( "sub %r2, %r4, %r3" != "subf 2, 3, 4" and "sub 2, 4, 3" != "subf 2, 3, 4" )

	[+] Line 73: 0x7c,0x43,0x20,0x51
		"sub. 2, 4, 3" != "subf. 2, 3, 4" ( "sub. 2, 4, 3" != "subf. 2, 3, 4" and "sub. 2, 4, 3" != "subf. 2, 3, 4" )

	[+] Line 79: 0x7c,0x43,0x20,0x10
		"subc %r2, %r4, %r3" != "subfc 2, 3, 4" ( "subc %r2, %r4, %r3" != "subfc 2, 3, 4" and "subc 2, 4, 3" != "subfc 2, 3, 4" )

	[+] Line 80: 0x7c,0x43,0x20,0x10
		"subc 2, 4, 3" != "subfc 2, 3, 4" ( "subc 2, 4, 3" != "subfc 2, 3, 4" and "subc 2, 4, 3" != "subfc 2, 3, 4" )

	[+] Line 128: 0x7c,0x43,0x21,0x5e
		"isel %r2, %r3, %r4, 4*cr1+gt" != "isel r2, r3, r4, cr1gt" ( "isel %r2, %r3, %r4, 4*cr1+gt" != "isel r2, r3, r4, cr1gt" and "isel 2, 3, 4, 5" != "isel r2, r3, r4, cr1gt" )



[-] Target: ppc64-encoding-fp.s.cs

[-] There are/is 10 failed test(s)
[-] Detailed report for ../MC/PowerPC/ppc64-encoding-ext.s.cs:

	[+] Line 50: 0x4e,0x80,0x00,0x20
		"bclr 0x14, 0, 0" != "blr" ( "bclr 20, 0, 0" != "blr" and "bclr 20, 0, 0" != "blr" )

	[+] Line 53: 0x4e,0x80,0x04,0x21
		"bcctrl 0x14, lt, 0" != "bctrl" ( "bcctrl 20, lt, 0" != "bctrl" and "bcctrl 20, 0, 0" != "bctrl" )

	[+] Line 394: 0x7c,0x44,0x18,0x50
		"sub %r2, %r3, %r4" != "subf 2, 4, 3" ( "sub %r2, %r3, %r4" != "subf 2, 4, 3" and "sub 2, 3, 4" != "subf 2, 4, 3" )

	[+] Line 395: 0x7c,0x44,0x18,0x51
		"sub. 2, 3, 4" != "subf. 2, 4, 3" ( "sub. 2, 3, 4" != "subf. 2, 4, 3" and "sub. 2, 3, 4" != "subf. 2, 4, 3" )

	[+] Line 396: 0x7c,0x44,0x18,0x10
		"subc 2, 3, 4" != "subfc 2, 4, 3" ( "subc 2, 3, 4" != "subfc 2, 4, 3" and "subc 2, 3, 4" != "subfc 2, 4, 3" )

	[+] Line 397: 0x7c,0x44,0x18,0x11
		"subc. 2, 3, 4" != "subfc. 2, 4, 3" ( "subc. 2, 3, 4" != "subfc. 2, 4, 3" and "subc. 2, 3, 4" != "subfc. 2, 4, 3" )

	[+] Line 532: 0x7c,0x62,0x1b,0x79
		"mr. %r2, %r3" != "or. 2, 3, 3" ( "mr. %r2, %r3" != "or. 2, 3, 3" and "mr. 2, 3" != "or. 2, 3, 3" )

	[+] Line 533: 0x7c,0x62,0x18,0xf8
		"not 2, 3" != "nor 2, 3, 3" ( "not 2, 3" != "nor 2, 3, 3" and "not 2, 3" != "nor 2, 3, 3" )

	[+] Line 534: 0x7c,0x62,0x18,0xf9
		"not. 2, 3" != "nor. 2, 3, 3" ( "not. 2, 3" != "nor. 2, 3, 3" and "not. 2, 3" != "nor. 2, 3, 3" )

	[+] Line 535: 0x7c,0x4f,0xf1,0x20
		"mtcr 2" != "mtcrf 255, 2" ( "mtcr 2" != "mtcrf 255, 2" and "mtcr 2" != "mtcrf 255, 2" )


[-] Target: ppc64-encoding-vmx.s.cs
```

Rot127 committed:
Remove manipulation of cs option during tests.

The only set of options which should be accepted, are the ones in the
test .cs files.

Rot127 committed:
Add NOREGNAME option to endocing tests.

Rot127 committed:
Revert "ppc: fix suite/MC/PowerPC/ppc64-encoding.s.cs"

This reverts commit 4b9d4857d7efce1ad0f45e3d48ae6463a5fe98ea.

With a consistent NOREGNAME option applied, these instructions
now do not emit the full register names.

Rot127 committed:
Fix PPC alias instruction tests.

Before the asm strings were non alias asm strings.
Alias are now better supported. Also the tests now match again the ones
in LLVM.

Rot127 committed:
Disable MSYNC feature.

Rot127 committed:
Fix book-II instruction tests - Sets instructions as in LLVM

Rot127 committed:
Add more ppc64 encoding tests.

Rot127 committed:
Add AIXOS feature to not enabled list.

Fixes mtspr and mfspr alias printing.

Rot127 commented:
Already at them.

Rot127 committed:
Add conditional branch alias which branch always.

Rot127 committed:
Add ppc64 encoding ext tests with new branch alias.

Rot127 committed:
Add Book-III tests from LLVM

Rot127 committed:
Remove wrong branch alias and add tests for them.

Conditional branches with BH != 0 have no alias.

Rot127 committed:
Add three more tests

Rot127 committed:
Add BOOKE as mode.

Rot127 committed:
Compile cstest with debug info

Rot127 commented:
Fix the detail tests tomorrow.

Rot127 committed:
Add option to print % sign in front of register names.

Rot127 committed:
Add support for S12 immediates of PS instructions.

Rot127 committed:
Add option to choose instruction detail set (between alias and real).

Rot127 committed:
Fix syntax/alias issuesin issues.cs

XVilka commented:
```
 Generating bindings for python
Error: Unable to convert ['ARMCC_CondCodes', 'cc;', '///<', 'conditional', 'code', 'for', 'this', 'insn']
Error: Unable to convert ['ARMVCC_VPTCodes', 'vcc;', '///<', 'Vector', 'conditional', 'code', 'for', 'this', 'instruction.']
Check test_basic.py ...
Traceback (most recent call last):
  File "/home/runner/work/capstone/capstone/bindings/python/./test_basic.py", line 133, in <module>
    test_class()
  File "/home/runner/work/capstone/capstone/bindings/python/./test_basic.py", line 122, in test_class
    print("0x%x:\t%s\t%s" % (insn.address, insn.mnemonic, insn.op_str))
                                           ^^^^^^^^^^^^^
  File "/home/runner/work/capstone/capstone/bindings/python/capstone/__init__.py", line 689, in mnemonic
    return self._raw.mnemonic.decode('ascii')
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UnicodeDecodeError: 'ascii' codec can't decode byte 0xd8 in position 0: ordinal not in range(128)
make: *** [Makefile:77: check] Error 1
FAILED
```

Rot127 committed:
fix false usage of arm details.

Rot127 committed:
Update python constants.

Rot127 committed:
Update python bindings with changes to PPC and cs_insn structs.

Rot127 committed:
Check for more edge cases when memory operands are emitted via printOperand()

Rot127 commented:
Finally got those bindings fixed. @kabeor @aquynh  Feel free to take a look.

XVilka commented:
@aquynh @kabeor ping

Rot127 commented:
@aquynh @kabeor Do you guys have already a time in mind when you can look at it?

kabeor commented:
@Rot127 I will start to review the code this weekend.

Rot127 committed:
Set correct values for the MCRegisterInfo struct.

XVilka commented:
> @Rot127 I will start to review the code this weekend.

Have you had a chance to look?

XVilka commented:
Would be nice to get this finally merged. It will reduce the amount of work and will make reviewing Aarch64 and Alpha PRs much easier.

kabeor commented:
Yeah I am reviewing the code, still need time.

XVilka commented:
Any updates?

XVilka commented:
@kabeor @aquynh ping

kabeor commented:
Every thing looks great! Merged.

XVilka commented:
Don't worry about red CI - it's a GitHub/network issue. Could be restarted later today:
```
Warning: Failed to download action 'https://api.github.com/repos/actions/checkout/tarball/f43a0e5ff2bd294095638e18286ca9a3d1956744'. Error: Response status code does not indicate success: 503 (Service Unavailable).
Warning: Back off 11.502 seconds before retry.
Warning: Failed to download action 'https://api.github.com/repos/actions/checkout/tarball/f43a0e5ff2bd294095638e18286ca9a3d1956744'. Error: Response status code does not indicate success: 503 (Service Unavailable).
Warning: Back off 27.394 seconds before retry.
```
https://www.githubstatus.com/incidents/c8zjb9351hlc

peace-maker commented on include/capstone/cs_operand.h:
```diff
@@ -24,15 +24,19 @@ typedef enum cs_op_type {
+	CS_AC_INVALID = 0,    ///< Uninitialized/invalid access type.
+	CS_AC_READ = 1 << 0,  ///< Operand read from memory or register.
+	CS_AC_WRITE = 1 << 1, ///< Operand write to memory or register.
+	CS_AC_READ_WRTE =
```
Typo wrte -> write
